h2o-algos/src/main/java/hex/gram/Gram.java

```diff
@@ -758,6 +758,34 @@ public void mul(double [] x, double [] res){
     }
   }
 
+  /*
+  This method will not allocate the extra memory and hence is considered for lowMemory systems.
+  However, need to consider case when you have categoricals!  Make them part of the matrix
+  in the multiplication process.
+   */
+  public void mul(double [] x, double [] res, boolean saveMemory){
+    if (saveMemory) {
+      int colSize = fullN();        // actual gram matrix size
+      int offsetForCat = colSize-_xx.length; // offset for categorical columns
+
+      for (int rowIndex = 0; rowIndex < colSize; rowIndex++) {
+        double d = 0;
+        for (int colIndex = 0; colIndex < rowIndex; colIndex++) {   // below diagonal
+          d += rowIndex>=offsetForCat?_xx[rowIndex-offsetForCat][colIndex] * x[colIndex]:0.0;
```

Comment:
what's the reason for adding 0.0? what's the reason for checking rowIndex n times? Can as well write 
if (rowIndex > offsetForCat) 
for (int colIndex = 0; colIndex < rowIndex; colIndex++) { 
 d += _xx[....][colIndex].